In [20]:
!pip install emoji wordsegment contractions

In [21]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import LabelEncoder
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score
import re
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import nltk
nltk.download('stopwords')

nltk.download('punkt')


from sklearn.pipeline import make_pipeline
import time


from sklearn.preprocessing import StandardScaler

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [22]:
import pandas as pd

reviews_df = pd.read_csv("processed_dataset.csv")

reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209708 entries, 0 to 209707
Data columns (total 11 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   review_id           209708 non-null  object
 1   user_id             209708 non-null  object
 2   business_id         209708 non-null  object
 3   stars               209708 non-null  int64 
 4   useful              209708 non-null  int64 
 5   funny               209708 non-null  int64 
 6   cool                209708 non-null  int64 
 7   text                209708 non-null  object
 8   date                209708 non-null  object
 9   sentiment           209708 non-null  int64 
 10  final_cleaned_text  209708 non-null  object
dtypes: int64(5), object(6)
memory usage: 17.6+ MB


In [23]:
reviews_df.head()

review_id                 user_id             business_id  \
0  J5Q1gH4ACCj6CtQG7Yom7g  56gL9KEJNHiSDUoyjk2o3Q  8yR12PNSMo6FBYx1u5KPlw   
1  HlXP79ecTquSVXmjM10QxQ  bAt9OUFX9ZRgGLCXG22UmA  pBNucviUkNsiqhJv5IFpjg   
2  JBBULrjyGx6vHto2osk_CQ  NRHPcLq2vGWqgqwVugSgnQ  8sf9kv6O4GgEb0j1o22N1g   
3  U9-43s8YUl6GWBFCpxUGEw  PAxc0qpqt5c2kA0rjDFFAg  XwepyB7KjJ-XGJf0vKc6Vg   
4  8T8EGa_4Cj12M6w8vRgUsQ  BqPR1Dp5Rb_QYs9_fz9RiA  prm5wvpp0OHJBlrvTj9uOg   

   stars  useful  funny  cool  \
0      2       1      0     0   
1      5       0      0     0   
2      5       0      0     0   
3      4       0      0     0   
4      5       0      0     0   

                                                text                 date  \
0  Went for lunch and found that my burger was me...  2018-04-04 21:09:53   
1  I needed a new tires for my wife's car. They h...  2020-05-24 12:22:14   
2  Jim Woltman who works at Goleta Honda is 5 sta...  2019-02-14 03:47:48   
3  Been here a few times to get some shrimp.  The...  2013-04-27 01:55:49   
4  This is one fantastic place to eat whether you...  2019-05-15 18:29:25   

   sentiment                                 final_cleaned_text  
0         -1  went lunch found burger meh obviou focu burger...  
1          1  need new tire wife car special order next day ...  
2          1  jim woltman work goleta honda 5 star knowledg ...  
3          1  time get shrimp got nice select differ fish pr...  
4          1  one fantast place eat whether hungri need good...

In [24]:
reviews_df.describe()

stars         useful          funny           cool  \
count  209708.000000  209708.000000  209708.000000  209708.000000   
mean        3.749294       1.172058       0.322901       0.497034   
std         1.478743       2.903865       1.435614       2.084638   
min         1.000000       0.000000       0.000000       0.000000   
25%         3.000000       0.000000       0.000000       0.000000   
50%         4.000000       0.000000       0.000000       0.000000   
75%         5.000000       1.000000       0.000000       0.000000   
max         5.000000     350.000000     116.000000     350.000000   

           sentiment  
count  209708.000000  
mean        0.439974  
std         0.840959  
min        -1.000000  
25%         0.000000  
50%         1.000000  
75%         1.000000  
max         1.000000

In [25]:
reviews_df.isnull().sum()

review_id             0
user_id               0
business_id           0
stars                 0
useful                0
funny                 0
cool                  0
text                  0
date                  0
sentiment             0
final_cleaned_text    0
dtype: int64

### Creating Target variable

In [26]:
# Creating the target variable (Positive, Negative, Neutral)
def label_review(stars):
    if stars >= 4:
        return '1'   # positive
    elif stars <= 2:
        return '-1'  # negative
    else:
        return '0'   # Neutral


reviews_df['sentiment'] = reviews_df['stars'].apply(label_review)

reviews_df.head()

review_id                 user_id             business_id  \
0  J5Q1gH4ACCj6CtQG7Yom7g  56gL9KEJNHiSDUoyjk2o3Q  8yR12PNSMo6FBYx1u5KPlw   
1  HlXP79ecTquSVXmjM10QxQ  bAt9OUFX9ZRgGLCXG22UmA  pBNucviUkNsiqhJv5IFpjg   
2  JBBULrjyGx6vHto2osk_CQ  NRHPcLq2vGWqgqwVugSgnQ  8sf9kv6O4GgEb0j1o22N1g   
3  U9-43s8YUl6GWBFCpxUGEw  PAxc0qpqt5c2kA0rjDFFAg  XwepyB7KjJ-XGJf0vKc6Vg   
4  8T8EGa_4Cj12M6w8vRgUsQ  BqPR1Dp5Rb_QYs9_fz9RiA  prm5wvpp0OHJBlrvTj9uOg   

   stars  useful  funny  cool  \
0      2       1      0     0   
1      5       0      0     0   
2      5       0      0     0   
3      4       0      0     0   
4      5       0      0     0   

                                                text                 date  \
0  Went for lunch and found that my burger was me...  2018-04-04 21:09:53   
1  I needed a new tires for my wife's car. They h...  2020-05-24 12:22:14   
2  Jim Woltman who works at Goleta Honda is 5 sta...  2019-02-14 03:47:48   
3  Been here a few times to get some shrimp.  The...  2013-04-27 01:55:49   
4  This is one fantastic place to eat whether you...  2019-05-15 18:29:25   

  sentiment                                 final_cleaned_text  
0        -1  went lunch found burger meh obviou focu burger...  
1         1  need new tire wife car special order next day ...  
2         1  jim woltman work goleta honda 5 star knowledg ...  
3         1  time get shrimp got nice select differ fish pr...  
4         1  one fantast place eat whether hungri need good...

In [27]:
reviews_df.shape

(209708, 11)

###Train Test Split

In [28]:
x_train, x_test, y_train, y_test = train_test_split(reviews_df['final_cleaned_text'], reviews_df['sentiment'], test_size=0.2, stratify=reviews_df['sentiment'], random_state=42)


print(y_train.value_counts())

sentiment
1     112467
-1     38654
0      16645
Name: count, dtype: int64


In [29]:
x_train.head()

181428    best pizza st pete opinion kind upset took lon...
32644     first thing need know pinewood social extrem i...
55173     food pretti good got meal grill beyond eat not...
204776    flowmast 44 put charger right bought figur bou...
73757     got number 11 omg sam made tast differ 30 year...
Name: final_cleaned_text, dtype: object

In [30]:
y_train.head()

181428     1
32644      1
55173     -1
204776     1
73757      1
Name: sentiment, dtype: object

###Vectorizing Text Data

In [31]:
count_vectorizer_unigram = CountVectorizer(ngram_range=(1, 1), max_features=5000, max_df=0.9,min_df=5)


x_train_count_unigram = count_vectorizer_unigram.fit_transform(x_train)
x_test_count_unigram = count_vectorizer_unigram.transform(x_test)

In [32]:
x_train_count_unigram.shape

(167766, 5000)

In [33]:
feature_names = count_vectorizer_unigram.get_feature_names_out()

x_train_count_unigram_df = pd.DataFrame(x_train_count_unigram.toarray(), columns=feature_names)


x_train_count_unigram_df.head()

10  100  1000  1010  1015  1030  10am  10pm  10th  11  ...  yr  yuck  yum  \
0   0    0     0     0     0     0     0     0     0   0  ...   0     0    0   
1   0    0     0     0     0     0     0     0     0   0  ...   0     0    0   
2   0    0     0     0     0     0     0     0     0   0  ...   0     0    0   
3   0    0     0     0     0     0     0     0     0   0  ...   0     0    0   
4   0    0     0     0     0     0     0     0     0   1  ...   0     0    0   

   yummi  yup  zero  zip  zone  zoo  zucchini  
0      0    0     0    0     0    0         0  
1      0    0     0    0     0    0         0  
2      0    0     0    0     0    0         0  
3      0    0     0    0     0    0         0  
4      0    0     0    0     0    0         0  

[5 rows x 5000 columns]

In [34]:
x_train.head()

181428    best pizza st pete opinion kind upset took lon...
32644     first thing need know pinewood social extrem i...
55173     food pretti good got meal grill beyond eat not...
204776    flowmast 44 put charger right bought figur bou...
73757     got number 11 omg sam made tast differ 30 year...
Name: final_cleaned_text, dtype: object

In [35]:
y_train.head()

181428     1
32644      1
55173     -1
204776     1
73757      1
Name: sentiment, dtype: object

###Sampling the data

Taking small samples of data to train and validate the SVM model

In [36]:
def sample_data(x_train, y_train, sample_size):


    indices = np.random.choice(x_train.shape[0], sample_size, replace=False)


    x_train_sample = x_train[indices]
    y_train_sample = y_train.iloc[indices] if hasattr(y_train, 'iloc') else y_train[indices]


    print(f"x_train_sample shape: {x_train_sample.shape}")
    print(f"y_train_sample shape: {y_train_sample.shape}")

    return x_train_sample, y_train_sample

Sample size taken = 30000

In [37]:
x_train_sample, y_train_sample = sample_data(x_train_count_unigram, y_train, sample_size=30000)

x_train_sample shape: (30000, 5000)
y_train_sample shape: (30000,)


Training and Cross-validating 30000 samples using **Stratified K-fold** with **k=5**

In [39]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.svm import SVC

# Create SVM model
svm_model = SVC(kernel='rbf', C=1, random_state=42)

# Create stratified k-fold cross-validation object
kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
scores = []
reports = []
matrices = []
start_time = time.time()
for train_index, val_index in kfold.split(x_train_sample, y_train_sample):
    X_train_fold, X_val_fold = x_train_sample[train_index], x_train_sample[val_index]
    y_train_fold, y_val_fold = y_train_sample.iloc[train_index], y_train_sample.iloc[val_index] # Use iloc to access data based on integer position
    svm_model.fit(X_train_fold, y_train_fold)
    y_pred = svm_model.predict(X_val_fold)
    scores.append(accuracy_score(y_val_fold, y_pred))
    reports.append(classification_report(y_val_fold, y_pred))
    matrices.append(confusion_matrix(y_val_fold, y_pred))
train_time = time.time() - start_time


print("Cross-validation scores:", scores)
print("Mean cross-validation score:", np.mean(scores))
print(f"Validation Time: {train_time:.2f} seconds for 30k samples")

# Print classification reports and confusion matrices for each fold
for i in range(len(reports)):
    print("Fold", i+1)
    print("Classification Report:")
    print(reports[i])
    print("Confusion Matrix:")
    print(matrices[i])

Cross-validation scores: [0.847, 0.8493333333333334, 0.841, 0.8445, 0.8488333333333333]
Mean cross-validation score: 0.8461333333333334
Validation Time: 1576.86 seconds for 30k samples
Fold 1
Classification Report:
              precision    recall  f1-score   support

          -1       0.83      0.79      0.81      1371
           0       0.56      0.09      0.15       577
           1       0.86      0.97      0.91      4052

    accuracy                           0.85      6000
   macro avg       0.75      0.62      0.62      6000
weighted avg       0.82      0.85      0.82      6000

Confusion Matrix:
[[1082   23  266]
 [ 136   50  391]
 [  86   16 3950]]
Fold 2
Classification Report:
              precision    recall  f1-score   support

          -1       0.81      0.81      0.81      1371
           0       0.54      0.07      0.13       577
           1       0.87      0.97      0.92      4052

    accuracy                           0.85      6000
   macro avg       0.74      

## Findings from Cross Validation with 30k Samples

**Cross-validation scores**: [0.847, 0.8493333333333334, 0.841, 0.8445, 0.8488333333333333]

**Mean cross-validation score**: 0.8461333333333334

**Validation Time**: 1576.86 seconds for 30k samples

### Fold 1

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.83      | 0.79   | 0.81     | 1371    |
| 0     | 0.56      | 0.09   | 0.15     | 577     |
| 1     | 0.86      | 0.97   | 0.91     | 4052    |

**Confusion Matrix**:
```
[[1082   23  266]
 [ 136   50  391]
 [  86   16 3950]]
```

### Fold 2

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.81      | 0.81   | 0.81     | 1371    |
| 0     | 0.54      | 0.07   | 0.13     | 577     |
| 1     | 0.87      | 0.97   | 0.92     | 4052    |

**Confusion Matrix**:
```
[[1108   26  237]
 [ 165   43  369]
 [  97   10 3945]]
```

### Fold 3

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.81      | 0.77   | 0.79     | 1371    |
| 0     | 0.64      | 0.09   | 0.16     | 577     |
| 1     | 0.85      | 0.97   | 0.91     | 4052    |

**Confusion Matrix**:
```
[[1055   16  300]
 [ 155   52  370]
 [ 100   13 3939]]
```

### Fold 4

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.82      | 0.80   | 0.81     | 1371    |
| 0     | 0.54      | 0.07   | 0.12     | 577     |
| 1     | 0.86      | 0.97   | 0.91     | 4052    |

**Confusion Matrix**:
```
[[1091   24  256]
 [ 131   40  406]
 [ 106   10 3936]]
```

### Fold 5

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.82      | 0.80   | 0.81     | 1371    |
| 0     | 0.62      | 0.08   | 0.14     | 577     |
| 1     | 0.87      | 0.97   | 0.92     | 4052    |

**Confusion Matrix**:
```
[[1113   25  233]
 [ 143   45  389]
 [  95   11 3946]]
```

Sample size taken = 40000

In [40]:
x_train_sample, y_train_sample = sample_data(x_train_count_unigram, y_train, sample_size=40000)

x_train_sample shape: (40000, 5000)
y_train_sample shape: (40000,)


Training and Cross-validating 40000 samples using **Stratified K-fold** with **k=5**

In [43]:
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.svm import SVC

# Create SVM model
svm_model = SVC(kernel='rbf', C=1, random_state=42)

# Create stratified k-fold cross-validation object
kfold = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation
scores = []
reports = []
matrices = []
start_time = time.time()
for train_index, val_index in kfold.split(x_train_sample, y_train_sample):
    X_train_fold, X_val_fold = x_train_sample[train_index], x_train_sample[val_index]
    y_train_fold, y_val_fold = y_train_sample.iloc[train_index], y_train_sample.iloc[val_index] # Use iloc to access data based on integer position
    svm_model.fit(X_train_fold, y_train_fold)
    y_pred = svm_model.predict(X_val_fold)
    scores.append(accuracy_score(y_val_fold, y_pred))
    reports.append(classification_report(y_val_fold, y_pred))
    matrices.append(confusion_matrix(y_val_fold, y_pred))
train_time = time.time() - start_time


print("Cross-validation scores:", scores)
print("Mean cross-validation score:", np.mean(scores))
print(f"Validation Time: {train_time:.2f} seconds for 40k samples")

# Print classification reports and confusion matrices for each fold
for i in range(len(reports)):
    print("Fold", i+1)
    print("Classification Report:")
    print(reports[i])
    print("Confusion Matrix:")
    print(matrices[i])

Cross-validation scores: [0.84975, 0.847375, 0.847375, 0.847375, 0.84525]
Mean cross-validation score: 0.847425
Validation Time: 2724.69 seconds for 40k samples
Fold 1
Classification Report:
              precision    recall  f1-score   support

          -1       0.83      0.80      0.82      1822
           0       0.54      0.11      0.18       795
           1       0.86      0.97      0.92      5383

    accuracy                           0.85      8000
   macro avg       0.74      0.63      0.64      8000
weighted avg       0.82      0.85      0.82      8000

Confusion Matrix:
[[1465   41  316]
 [ 203   87  505]
 [ 105   32 5246]]
Fold 2
Classification Report:
              precision    recall  f1-score   support

          -1       0.83      0.80      0.81      1822
           0       0.60      0.11      0.18       795
           1       0.86      0.97      0.91      5383

    accuracy                           0.85      8000
   macro avg       0.76      0.63      0.64      8000

## Findings from Cross Validation with 40k Samples

**Cross-validation scores**: [0.84975, 0.847375, 0.847375, 0.847375, 0.84525]

**Mean cross-validation score**: 0.847425

**Validation Time**: 2724.69 seconds for 40k samples

### Fold 1

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.83      | 0.80   | 0.82     | 1822    |
| 0     | 0.54      | 0.11   | 0.18     | 795     |
| 1     | 0.86      | 0.97   | 0.92     | 5383    |

**Confusion Matrix**:
```
[[1465   41  316]
 [ 203   87  505]
 [ 105   32 5246]]
```

### Fold 2

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.83      | 0.80   | 0.81     | 1822    |
| 0     | 0.60      | 0.11   | 0.18     | 795     |
| 1     | 0.86      | 0.97   | 0.92     | 5383    |

**Confusion Matrix**:
```
[[1462   31  329]
 [ 179   84  532]
 [ 125   25 5233]]
```

### Fold 3

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.83      | 0.79   | 0.81     | 1822    |
| 0     | 0.59      | 0.11   | 0.18     | 795     |
| 1     | 0.86      | 0.98   | 0.92     | 5383    |

**Confusion Matrix**:
```
[[1438   40  344]
 [ 185   86  524]
 [ 107   21 5255]]
```

### Fold 4

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.82      | 0.79   | 0.80     | 1823    |
| 0     | 0.59      | 0.10   | 0.18     | 795     |
| 1     | 0.86      | 0.97   | 0.92     | 5382    |

**Confusion Matrix**:
```
[[1449   37  337]
 [ 220   83  492]
 [ 115   20 5247]]
```

### Fold 5

**Classification Report**:

| Class | Precision | Recall | F1-score | Support |
|-------|-----------|--------|----------|---------|
| -1    | 0.82      | 0.79   | 0.81     | 1823    |
| 0     | 0.60      | 0.11   | 0.18     | 795     |
| 1     | 0.87      | 0.98   | 0.92     | 5382    |

**Confusion Matrix**:
```
[[1455   35  333]
 [ 192   85  518]
 [ 110   23 5249]]
```

##Conclusion

The code performs exhaustive testing of the performance of the SVM classifier on 5-fold cross-validation with two samples of size 30,000 and 40,000 of the dataset. The code will train and evaluate the SVM model using StratifiedKFold Cross Validation, and it makes use of the grid search functionality for tuning hyperparameters. Detailed performance metrics on each fold are extracted for precision, recall, F1-score per class, confusion matrices. For both of these, I save cross-validation scores, the mean score, and the time taken to validate. The results are quite consistent for different folds. It seems that the model did an excellent job in classes -1 and 1, with class 0 still having rather low precision and recall. The larger dataset of 40k samples thus results in a slight improvement in the mean cross-validation score to 0.847425, as against 0.8461333333333334 for the 30k dataset. According to the confusion matrices, misclassifications tend to happen most between classes 0 and 1, showing where a model might be improved or features could be contrived. Validation time increases from 1576.86 seconds to 2724.69 seconds.